### PR TITLE
Add peer validation to ALTS implementation.

### PR DESCRIPTION
**Pull request by yangminzhu**
**What this PR does / why we need it:**

- Added the peer validation to ALTS implementation.
- Manually tested with custom config and environment on GKE.

**Special notes for your reviewer:**
None

**Release note:**
None

### CODE REVIEW

**Comment by JimmyCYJ** [comment]
Please fix format error in proxy-presubmit test. You can run ./script/check-style to fix that.
And add PR description back. e.g. https://github.com/istio/proxy/pull/1019

File: src/envoy/alts/tsi_transport_socket.h [hunk 1]
```diff
@@ -99,13 +106,15 @@ class TsiSocket : public Network::TransportSocket,
  */
 class TsiSocketFactory : public Network::TransportSocketFactory {
  public:
-  explicit TsiSocketFactory(HandshakerFactory handshaker_factory);
+  explicit TsiSocketFactory(HandshakerFactory handshaker_factory,
+                            HandshakeValidator handshake_validator);
```
Comment (JimmyCYJ) [inline src/envoy/alts/tsi_transport_socket.h]: I think explicit is not needed anymore, as there are more than one argument.

File: src/envoy/alts/tsi_transport_socket.h [hunk 2]
```diff
@@ -35,7 +37,8 @@ class TsiSocket : public Network::TransportSocket,
   /**
    * @param handshaker_factory a function to initiate a TsiHandshaker
    */
-  explicit TsiSocket(HandshakerFactory handshaker_factory);
+  explicit TsiSocket(HandshakerFactory handshaker_factory,
+                     HandshakeValidator handshake_validator);
```
Comment (JimmyCYJ) [inline src/envoy/alts/tsi_transport_socket.h]: I think explicit is not needed anymore.

File: src/envoy/alts/tsi_transport_socket.cc [hunk 3]
```diff
@@ -82,14 +85,23 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(
...
                    callbacks_->connection(), peer.property_count);
-    // TODO(lizan): validate peers
     for (size_t i = 0; i < peer.property_count; ++i) {
       ENVOY_CONN_LOG(debug, "  {}: {}", callbacks_->connection(),
                      peer.properties[i].name,
                      std::string(peer.properties[i].value.data,
                                  peer.properties[i].value.length));
     }
+    std::string err = "";
+    bool peer_validated = handshake_validator_(peer, err);
```
Comment (JimmyCYJ) [inline src/envoy/alts/tsi_transport_socket.cc]: Do we need to check if handshake_validator_ is nullptr?

File: src/envoy/alts/alts_socket_factory.cc [hunk 4]
```diff
@@ -29,6 +29,36 @@ namespace Envoy {
...
+      for (int j = 0; j < targets.size(); ++j) {
+        if (value.compare(targets[j]) == 0) {
+          return true;
+        }
+      }
+    }
+  }
+
+  err = "Couldn't find peer's service account in targets: ";
+  for (int i = 0; i < targets.size(); ++i) {
```
Comment (qiwzhang) [inline src/envoy/alts/alts_socket_factory.cc]: For easy debugging, it is better to log both peers and targets.

File: src/envoy/alts/tsi_transport_socket.h (hunk 1 above)
Comment (yangminzhu) [reply src/envoy/alts/tsi_transport_socket.h]: Done.

File: src/envoy/alts/tsi_transport_socket.h (hunk 2 above)
Comment (yangminzhu) [reply src/envoy/alts/tsi_transport_socket.h]: Thanks for the review, also updated the comments.

File: src/envoy/alts/tsi_transport_socket.cc (hunk 3 above)
Comment (yangminzhu) [reply src/envoy/alts/tsi_transport_socket.cc]: I'm assuming it's used the same as handshake_factory which should always be provided by the caller. If the caller do not want to validate peer, they can pass in a lambda that always return true.

File: src/envoy/alts/alts_socket_factory.cc (hunk 4 above)
Comment (yangminzhu) [reply src/envoy/alts/alts_socket_factory.cc]: The peers will be logged in src/envoy/alts/tsi_transport_socket.cc just before the handshake_validator is called.

File: src/envoy/alts/tsi_transport_socket.cc (hunk 3 above)
Comment (yangminzhu) [reply src/envoy/alts/tsi_transport_socket.cc]: After a second thought, I think it doesn't harm to add such check, so updated accordingly.

File: src/envoy/alts/tsi_transport_socket.h [hunk 5]
```diff
@@ -24,6 +24,8 @@ namespace Envoy {
 namespace Security {
 
 typedef std::function<TsiHandshakerPtr(Event::Dispatcher&)> HandshakerFactory;
+typedef std::function<bool(const tsi_peer& peer, std::string& err)>
```
Comment (lizan) [inline src/envoy/alts/tsi_transport_socket.h]: Add a comment for how they are used. err is an output param, right?

File: src/envoy/alts/tsi_transport_socket.cc [hunk 6]
```diff
@@ -82,13 +85,28 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(
...
                    callbacks_->connection(), peer.property_count);
-    // TODO(lizan): validate peers
     for (size_t i = 0; i < peer.property_count; ++i) {
       ENVOY_CONN_LOG(debug, "  {}: {}", callbacks_->connection(),
                      peer.properties[i].name,
                      std::string(peer.properties[i].value.data,
                                  peer.properties[i].value.length));
     }
+    if (handshake_validator_) {
+      std::string err = "";
```
Comment (lizan) [inline src/envoy/alts/tsi_transport_socket.cc]: no need `= ""`

File: src/envoy/alts/alts_socket_factory.cc [hunk 7]
```diff
@@ -29,6 +29,36 @@ namespace Envoy {
 namespace Server {
 namespace Configuration {
 
+using ::google::protobuf::RepeatedPtrField;
+
+// Returns true if the peer's service account is found in targets, otherwise
+// returns false and fills out err with an error message.
+static bool doValidate(const tsi_peer &peer,
+                       const RepeatedPtrField<std::string> &targets,
```
Comment (lizan) [inline src/envoy/alts/alts_socket_factory.cc]: pre-process `targets` to a `unordered_set` then you don't have to do O(N^2) validation below.

File: src/envoy/alts/example.yaml [hunk 8]
```diff
@@ -19,6 +19,9 @@ static_resources:
         name: alts
         config:
           handshaker_service: "169.254.169.254:8080"
+          # Please add the actual service account used in your environment, otherwise the ALTS
+          # connection will be closed due to validation failure.
+          target_service_accounts: ["test-service-account"]
```
Comment (lizan) [inline src/envoy/alts/example.yaml]: Can you rename `target_service_accounts` to `peer_service_accountss`? Thanks!

**Pull request merged**